First pass of adjustments

encukou opened the pull request:
See individual commits

encukou committed:
Put the 8 in UTF-8; it's not UTF-16. Use size.

Size can be omitted for identifiers and human-readable messages?

encukou committed:
Use int64 in FromInt64

encukou committed:
Remove _s specializations

Users should pre-create interned constant strings and stash them in
"global" storage.

encukou committed:
Correct arg types for Exception.New/New_s

encukou committed:
Handle Str name collision

encukou committed:
Use a valid signature for CFunctionPointer

encukou committed:
Make this compile

- include stdint.h
- define PyRef, PyContext, Utf8String
- typedefs first
- append an underscore to `default`
- make the names match

encukou committed:
Regenerate PyAPI.h

encukou committed:
Delete unused class

encukou committed:
Remove more _s specializations

encukou commented:
Regarding the `_s` specializations:
- For static strings, I think we should encourage creating “global” objects in the module state. We'll need to make such “global” objects accessible anyway.
- If you disagree let's argue, but let's not add these *right* now?

I left a few of them (like `PyNI_Exception_RaiseNew_s`, which you'd commonly use with a `sprintf`-like) so we can figure out how they should work in general.

markshannon commented on PyAPI.h:
```diff
@@ -12,15 +16,53 @@ int PyRef_IsMissing(PyRef ref);
+/* typedefs for generate.types_and_modules */
+typedef PyRef (*PyNI_VectorCall_FuncPtr)(PyContext ctx, PyRef callable, PyRef args[], intptr_t nargsf, PyTupleRef kwnames);
+
+typedef int32_t CType;
```
This declared as an enum later, so there will be a type conflict.

It's OK to declare it as an enum, as it is never used as a parameter or return value.

markshannon commented on PyAPI.h:
```diff
@@ -424,11 +467,12 @@ if (PyNI_CanDownCastToBytes(REF)) {\
 int PyNI_StrBuilder_AppendStr(PyContext ctx, PyRef Self, PyStrRef s);
 
-int PyNI_StrBuilder_AppendUtf8String(PyContext ctx, PyRef Self, const UtfString s);
+int PyNI_StrBuilder_AppendUtf8String(PyContext ctx, PyRef Self, const Utf8String s, uintptr_t size);
```
This is an odd one out. It takes a length as well as a string.
Presumably for performance to avoid scanning the string twice?

Might be worth adding a comment somewhere.

encukou replied:
`FromUtf8String` is the same.

The idea is that *identifiers* and *human-readable messages* [edit: and *source code* and *filenames*] are NUL-terminated, while `str` contents can contain embedded NUL and are passed as buffer & size.

I guess it's worth a `Utf8StringZ` type. And perhaps extra functions like `PyNI_StrBuilder_AppendUtf8StringZ`

markshannon replied:
Yes, we need to know the length of the string.
`Utf8String_Z` (`NULL` terminated strings) have a length, but `Utf8String` will need one.

Can we use just `Utf8String` for a string, and convert that to a `char *str, uintptr_t len` pair in the code generator, or whatever is appropriate for Rust or C++.

markshannon commented on PyAPI.h:
```diff
@@ -666,31 +681,23 @@ if (PyNI_CanDownCastToFrameStack(REF)) {\
+};
 
-PyCodeRef PyNI_Eval_Compile(PyContext ctx, PyRef src, PyRef filename, PySourceKindRef kind);
+PyCodeRef PyNI_Eval_Compile(PyContext ctx, PyRef src, PyRef filename, EvalSourceKind kind);
```
Until we have C23, I think `kind` will need to be an `int`.

markshannon commented on PyAPI.h:
```diff
@@ -706,60 +713,58 @@ typedef enum CType {
-typedef PyRef (*PyNI_Getter_FuncPtr)(PyContext ctx, PyRef obj, PyFieldDescriptorRef field, void* data);
-
-typedef void (*PyNI_Setter_FuncPtr)(PyContext ctx, PyRef obj, PyFieldDescriptorRef field, void* data, PyRef value);
+/* (<class 'generate.types_and_modules.FieldDescriptor'>, '') */
```
Either make this comment clearer (like "generated from types_and_modules.FieldDescriptor"), or remove it.

markshannon commented on generate/__main__.py:
```diff
@@ -106,38 +122,62 @@ def generate_api_func(func, namespace):
+        print(f"}};")
     elif issubclass(cls, Struct):
-        print(f"typedef struct {namespace}{cls.__name__} {{")
+        print(f'/* {cls, namespace} */')
```
```suggestion
        print(f'/* Generated from {cls.__qualname__} */')
```

markshannon reviewed (changes requested):
Looks good overall.

I think there is a conflict in the types for `CTypes` and I have a few other quibbles.

encukou committed:
Revert enum from int32_t

encukou committed:
Zero-terminated string

encukou committed:
Pass enum arguments as int

encukou committed:
Fix generated comment

encukou committed:
Regenerate the header

encukou committed:
More utf_string_z

encukou commented:
Does this look better?

markshannon commented on PyAPI.h:
```diff
@@ -468,32 +514,20 @@ PyRef PyNI_Object_GetItem(PyContext ctx, PyRef obj, PyRef key);
 
 PyRef PyNI_Object_GetItem_i(PyContext ctx, PyRef obj, intptr_t key);
 
-PyRef PyNI_Object_GetItem_s(PyContext ctx, PyRef obj, const UtfString key);
```
Should we add a `_z` version instead?
```C
PyRef PyNI_Object_GetItem_z(PyContext ctx, PyRef obj, const char *key);
```

markshannon reviewed (commented):
Looks good.
A couple of questions/suggestions about using C strings.

Should we declare a `Utf8String` as a single item in the declaration and generate a `const char *, uintptr_t` pair of arguments when generating the C, or should we require an explicit length parameter in the declaration?
Or should we, possibly with an eye on future Rust implementation, pass a string as a struct composed of a pointer and a length?

encukou committed:
Skip entries defined in other modules.

The generator broke on Python 3.14 (which has PEP 649 – Deferred
Evaluation Of Annotations Using Descriptors) due to `Str` being
deleted from `generate.vm` globals when annotations are evaluated.

Don't delete these. Instead, skip entries that aren't defined in the
module we're processing.

This removes a duplicate `PyNI_VectorCall_FuncPtr` typedef; otherwise
the result is the same.

encukou commented:
> Should we declare a Utf8String as a single item in the declaration and generate a const char *, uintptr_t pair of arguments when generating the C, or should we require an explicit length parameter in the declaration?

That sounds great! It might be a bit of a rabbit hole though; would you be OK with leaving it to the next pass?

> possibly with an eye on future Rust implementation, pass a string as a struct composed of a pointer and a length?

I'd rather have two arguments in the ABI.

AFAIK, ABIs for passing structs by value tends to be a bit complicated if your compiler or ffi library doesn't have it all figured out. And I don't think it would buy us much over simply having two register-sized arguments.

Also AFAIK, Rust doesn't guarantee much about memory layouts, so that wrapper would need to unpack things (or rely on internals).
